cpp/test/core/apply_tests.cpp

```diff
@@ -134,6 +141,210 @@ Bonded(uint16_t version = bond::v1)
+    {}
+
+    void End() const
+    {
+        if (!_structs)
+            return;
+
+        _this << "};\n";
+        _structs->emplace(*_qualifiedName, _this.str());
+    }
+
+private:
+    template <typename T>
+    const std::string& StructName(const T& value) const
+    {
+        IDLBuilder that{ *_structs };
+        Apply(that, value);
+        return *that._qualifiedName;
+    }
+
+    template <typename T, typename Reader, typename boost::enable_if<bond::is_basic_type<T> >::type* = nullptr>
+    void TypeName(const bond::value<T, Reader>& /*value*/) const
+    {
+        _this << bond::detail::type<typename std::conditional<std::is_enum<T>::value, std::int32_t, T>::type>::name();
+    }
+
+    template <typename T>
+    void TypeName(const T& value) const
+    {
+        _container = bond::GetTypeId(value);
+
+        if (_container == bond::BT_STRUCT)
+            _this << StructName(value);
+        else
+            Apply(*this, value);
+    }
+
+    template <typename... T>
+    void Container(std::size_t size, const T&... value) const
+    {
+        BOOST_VERIFY(size == 0);
+        assert(bond::BT_LIST <= _container && _container <= bond::BT_MAP);
+        assert((sizeof...(T) == 2) == (_container == bond::BT_MAP));
+
+        const char* labels[] = { "list", "set", "map" };
+        _this << labels[_container - bond::BT_LIST] << '<';
+        (void)std::initializer_list<int>{ (TypeName(value), _this << ", ", 0)... };
+        _this.seekp(-2, std::ios_base::end);
+        _this << '>';
+    }
+
+    template <typename T, typename Enable = void> struct
+    default_value_field;
+
+    template <typename T> struct
+    default_value_field<T, typename boost::enable_if<std::is_unsigned<T> >::type> : std::integral_constant<std::uint64_t bond::Variant::*, &bond::Variant::uint_value> {};
+
+    template <typename T> struct
+    default_value_field<T, typename boost::enable_if<bond::is_signed_int_or_enum<T> >::type> : std::integral_constant<std::int64_t bond::Variant::*, &bond::Variant::int_value> {};
+
+    template <typename T> struct
+    default_value_field<T, typename boost::enable_if<std::is_floating_point<T> >::type> : std::integral_constant<double bond::Variant::*, &bond::Variant::double_value> {};
+
+    template <typename T> struct
+    default_value_field<T, typename boost::enable_if<bond::is_string<T> >::type> : std::integral_constant<std::string bond::Variant::*, &bond::Variant::string_value> {};
+
+    template <typename T> struct
+    default_value_field<T, typename boost::enable_if<bond::is_wstring<T> >::type> : std::integral_constant<std::wstring bond::Variant::*, &bond::Variant::wstring_value> {};
+
+    template <typename T>
+    void FormatValue(const T& value) const
+    {
+        _this << value;
+    }
+
+    void FormatValue(const std::wstring& value) const
+    {
+        std::transform(value.begin(), value.end(), std::ostreambuf_iterator<char>(_this), [](wchar_t wc) { return char(wc); });
```

Comment:
I guess none of the default string values in the test cases use non-ASCII values? ;-)

Reply:
Yeah, none :) We could replace all these to use wide strings to support both.